### DISCUSSION TITLE
shared class ->multiple pathway for object oriented constructions (#Explained / #SharedType / #Engagement)

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5508] [Language Ideas]
### TL;DR

TL;DR :Instead of chaining functions together we chain objects together, not through single inheritance but multiple, not just declare class in another class but in a way more visible.

We can combine multiple interface, but you have to implement stuffs directly in your class manually, which makes composition more verbose and explicit defeating the purpose of orchestration (to draw complexity in simple forms for aesthetic pleasing experience to let developers be more happy which in turns inspire what they are truly capable of)

So C# can feel more like having a functional approach to things without really going against its narrative because it can use what's already there it is just syntax sugar.

### Background and motivation

We currently don't have a way to combine multiple type into a single type. (We can combine multiple interface)

### API Proposal

Let the compile and IDE recognize the synctax then it will be fine, the class just inheirts all shared class in the background like you would inheirt one which is what we're doing.

### API Usage

```C#
public shared class A
{
    public int e1=0;
    public string e2;
}
public shared class B
{
    public float e2;
}
public shared class C
{
    public B e3;
    public int e4;
}
public class D:A,B,C
{
    public shared D()
    {
        pathway (instance):
        //Pretend someone passed instances of class A, B, C
case A://Refer to by class name
        return new { _=> instance.e1,
                           _=>instance.e2 };//anonymous type for infered name
case B:
        return new { _=>instance.e2};
case C:
        return new { _=instance.e3};//e4 ignored
        default:
        //resolve conflicts here, so when D is constructed from one class it's ok, but because type composition (you can build object of class D by building it with A first, then B, then C etc.) is allowed when D is composed with multiple types the members that's inherited but is conflict with eachother should be resolved.
        B->e2=>A->e2; //e2 from B overides e2 from A when both classes are used to compose D in the past or in the future
    }
}//D has all members from A,B,C, constructor makes sure any conflicts between combined classes are resolved
```

### Alternative Designs

Just another idea, if an object is a composite class of A & B, then this class should be able to be used to compose D as long as D implemented / inherited the two shared class.

Because .NET remembers the type of the object, even when you cast the object to { object } type, shared classes for now should follow the type safety we have, that means though you can somehow using a new syntax to create class D from A you can't cast A to D.

If class D didn't implement a pathway to construct from class A (case A:), a warning perhaps would be suitable, while this scenario means A is ignored when used / present in the composing / construction of the class / shared class (naming unsettled).

**Comment by Xyncgas** [comment]
The benifit of this is you can construct an object step by step, I still havent figured out the syntax yet, but the idea is there which I think is powerful because one of the powerful thing from functional programming that we could get was its ability to compose at compile time what can happen and kind of link the things together, with shared class, you can do this now, which the way I see it is just simply one class having all three classes' instance, at least it's one way to implement it, which is simple enough, which then it's synctatic sugars added on top of it to make members of all three instances all available as if they're one class.

Type casting is going to work imo.

One thing I am worried is when the types are having members of the same name and we ended up hiding them or whatever, then the class isn't really 3 complete class acting as one, and we need to figure out how to serialize to it. but we will always have the constructor / pathway as a map, which holds the definition of how the class was composed of the 3 (or others.) classes, from there we know which members of the current composed class is belonged of the shared classes.

**Reply by Xyncgas** [reply]
TL;DR :Instead of chaining functions together we chain objects together, not through single inheritance but multiple, not just declare class in another class but in a way more visible.

So C# can feel more like having a functional approach to things without really going against its narrative because it can use what's already there it is just syntax sugar.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/1212

This feels like a crack at MI but with much more complicated syntax.  You still hand-wave away the problem of overlapping state, which is definitely the hardest thing that would need to be addressed.  You don't address how this would work in the runtime either, or how these types can possibly span assemblies.

**Reply by Xyncgas** [reply]
c++ has multiple inheritance already

**Reply by Xyncgas** [reply]
my insight lies in that c# shall be unique from c++, but in a way we can basically do this by simply hide members from IDE & in the compiler (that are overlapping, but resolved), then trim away unreferenced members later.

**Reply by Xyncgas** [reply]
how's that sound for u

**Reply by Xyncgas** [reply]
to elaborate a bit further, I am thinking I am taking three classes A,B,C, now we flatten them and all their members are visible to us, we can do this by putting A,B,C inside D like we normally do, by declaring them inside the class, or we can just add multiple inheritance and they are now visible, the overlapping members will be resolved before compiling by users writing the code in the example syntax to throw away / over power members that are conflicting.

**Reply by HaloFour** [reply]
> c++ has multiple inheritance already

C++ does have multiple inheritance already and the problems with it are well known, which is why .NET and C# has explicitly decided to avoid it.

> but in a way we can basically do this by simply hide members from IDE & in the compiler (that are overlapping, but resolved), then trim away unreferenced members later.

That doesn't describe whether or not the state _should_ be shared, nor does it describe what to do in the case of diamond inheritance of state.  It also doesn't describe how this works in the runtime.  The compiler can't emulate this.

The language team has been pretty opposed to the notion of exploring MI any time it's been brought up in the past.

**Reply by Xyncgas** [reply]
With shared functions, we are introducing object oriented constructions / object oriented compositions, just like functional programming that we can compose functions and make new functions, we can stand on our object oriented language and open a whole new world of possibilities to compose types. Although I can't tell you all the new ways for us to do things, I certainly do believe in its beauty.

**Reply by Xyncgas** [reply]
Ok, thanks for the input on this. I will keep thinking and if I come up with something good I will let you know.

**Reply by Xyncgas** [reply]
to address the issue about a shared state, I don't think there is shared state, because it's just a template behind the curtain. You declare how this object can be constructed by multiple classes, it then shall be constructed. for one the diamond problem can be easily turned in a diamond principle, such that :When in conflict, look at the target class being composed for information regarding what member gets to overwrite another, if the conflicting members totally belong to the same base class than just have one base class instead of two, you can cast up and down all you want, well, maybe not down casting maybe we don't allow it for now, but I think this is acceptable

**Reply by HaloFour** [reply]
If it's just a template and you're looking to have the class delegate to several other instances without actually inheriting from them then I would suggest that this proposal could be accomplished through source generators.

**Comment by TonyValenti** [comment]
Here's what you do:
1.  Create an interface for each set of properties that are logically grouped together (we'll call this a property group interface). 
2.  Create classes that implement the property group interfaces - one interface per class(we'll call these property group implementations). 
3.  For your "aggregate" classes, create private fields that are of the property group interface and assigned to a new instance of a property group implementation.  Your aggregate class should implement all interfaces and redirect all calls to the appropriate property group implementor.
